README.md
# replitcode-local README

replitcode-local is an extension to work with replit-code-v1-3b LLM running on Flask.


CHANGELOG.md
# Change Log

All notable changes to the "replitcode-local" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
import * as vscode from 'vscode';
import axios from 'axios';

let typingTimer: NodeJS.Timeout | null = null;
let typingDelay: number = vscode.workspace.getConfiguration().get('replitcode-local.typingDelay') || 1000;
let flaskUrl: string = vscode.workspace.getConfiguration().get('replitcode-local.flaskUrl') || 'http://localhost:5000/predict';
let codeLines: number = (vscode.workspace.getConfiguration().get<number>('replitcode-local.codeLines') || 3) - 1;

async function getCodeFromFlask(code: string): Promise<string> {
  try {
    const response = await axios.post(flaskUrl, { "input_data": code });
    console.log('Done request, answer:', response.data.generated_code);
    return response.data.generated_code;
  } catch (error) {
    console.error('Error fetching code from Flask:', error);
    return '';
  }
}

export function activate(context: vscode.ExtensionContext) {
  // Register a configuration change listener
  vscode.workspace.onDidChangeConfiguration(handleConfigurationChange);

  console.log('Done activate');

  const provider: vscode.CompletionItemProvider = {
    // @ts-ignore
    provideInlineCompletionItems: async (document, position, context, token) => {
      const startLine = Math.max(position.line - codeLines, 0);
      const textBeforeCursor = document.getText(
        new vscode.Range(position.with(startLine, 0), position)
      );
      console.log('textBeforeCursor', textBeforeCursor);

      let items: vscode.CompletionItem[] = [];

      if (textBeforeCursor) {
        if (typingTimer) {
          clearTimeout(typingTimer);
        }

        return new Promise((resolve) => {
          typingTimer = setTimeout(async () => {
            try {
              const genCode = await getCodeFromFlask(textBeforeCursor);
              console.log('genCode', genCode);
              if (genCode) {
                const completionItem = new vscode.CompletionItem(genCode);
                completionItem.range = new vscode.Range(position.translate(0, genCode.length), position);
                completionItem.insertText = genCode;
                items.push(completionItem);
              }
            } catch (err: any) {
              vscode.window.showErrorMessage(err.toString());
            }

            resolve({ items });
          }, typingDelay);
        });
      }

      return { items };
    },
  };
  // @ts-ignore
  vscode.languages.registerInlineCompletionItemProvider({ pattern: '**' }, provider);
}

function handleConfigurationChange(event: vscode.ConfigurationChangeEvent) {
  if (event.affectsConfiguration('replitcode-local.typingDelay')) {
    typingDelay = vscode.workspace.getConfiguration().get('replitcode-local.typingDelay') || 1000;
  }

  if (event.affectsConfiguration('replitcode-local.flaskUrl')) {
    flaskUrl = vscode.workspace.getConfiguration().get('replitcode-local.flaskUrl') || 'http://localhost:5000/predict';
  }

  if (event.affectsConfiguration('replitcode-local.codeLines')) {
    const newcodeLines = vscode.workspace.getConfiguration().get('replitcode-local.codeLines') as number;
    codeLines =  (newcodeLines || 1000) - 1;
  }
}

export function deactivate() {
  // Clean up resources if needed
}


src/old_extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "replitcode-local" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json
	let disposable = vscode.commands.registerCommand('replitcode-local.helloWorld', () => {
		// The code you place here will be executed every time your command is executed
		// Display a message box to the user
		vscode.window.showInformationMessage('Hello World from ReplitCode Local!');
	});

	context.subscriptions.push(disposable);
}

// This method is called when your extension is deactivated
export function deactivate() {}


src/work_extension.ts
import * as vscode from 'vscode';
import axios from 'axios';


async function getCodeFromFlask(code: string): Promise<string> {
    try {
      const response = await axios.post('http://localhost:5000/predict', { "input_data": code });
      console.log('Done reauest, aswer', response.data.generated_code);
      return response.data.generated_code;
    } catch (error) {
      console.error('Error fetching code from Flask:', error);
      return '';
    }
  }


export function activate(_: vscode.ExtensionContext) {

    const provider: vscode.CompletionItemProvider = {
        // @ts-ignore
        provideInlineCompletionItems: async (document, position, context, token) => {
            
            const startLine = Math.max(position.line - 2, 0);
            const textBeforeCursor = document.getText(
                new vscode.Range(position.with(startLine, 0), position)
            );
            console.log('textBeforeCursor', textBeforeCursor);

            let items: any[] = [];

            if (textBeforeCursor) {
                let genCode;
                try {
                    genCode = await getCodeFromFlask(textBeforeCursor);
                    console.log('genCode', genCode);
                    if (genCode) {
                        return [{
                            text: genCode,
                            insertText: genCode,
                            range: new vscode.Range(position.translate(0, genCode.length), position)
                        }];
                    };
                    
                } catch (err: any) {
                    vscode.window.showErrorMessage(err.toString());
                }
            }
            return {items};
        },
    };

    // @ts-ignore
    vscode.languages.registerInlineCompletionItemProvider({ pattern: '**' }, provider);
}

src/work1_extension.ts
import * as vscode from 'vscode';
import axios from 'axios';

let typingTimer: NodeJS.Timeout | null = null;
let typingDelay: number;
let flaskUrl: string;

async function getCodeFromFlask(code: string): Promise<string> {
  try {
    const response = await axios.post(flaskUrl, { "input_data": code });
    console.log('Done request, answer:', response.data.generated_code);
    return response.data.generated_code;
  } catch (error) {
    console.error('Error fetching code from Flask:', error);
    return '';
  }
}

export function activate(_: vscode.ExtensionContext) {
  const config = vscode.workspace.getConfiguration('replitcode-local');
  typingDelay = config.get('typingDelay', 1000);
  flaskUrl = config.get('flaskUrl', 'http://localhost:5000/predict');

  console.log('activate done');

  const provider: vscode.CompletionItemProvider = {
    // @ts-ignore
    provideInlineCompletionItems: async (document, position, context, token) => {
      const startLine = Math.max(position.line - 2, 0);
      const textBeforeCursor = document.getText(
        new vscode.Range(position.with(startLine, 0), position)
      );
      console.log('textBeforeCursor', textBeforeCursor);

      let items: vscode.CompletionItem[] = [];

      if (textBeforeCursor) {
        if (typingTimer) {
          clearTimeout(typingTimer);
        }

        return new Promise((resolve) => {
          typingTimer = setTimeout(async () => {
            try {
              const genCode = await getCodeFromFlask(textBeforeCursor);
              console.log('genCode', genCode);
              if (genCode) {
                const completionItem = new vscode.CompletionItem(genCode);
                completionItem.range = new vscode.Range(position.translate(0, genCode.length), position);
                completionItem.insertText = genCode;
                items.push(completionItem);
              }
            } catch (err: any) {
              vscode.window.showErrorMessage(err.toString());
            }

            resolve({ items });
          }, typingDelay);
        });
      }

      return { items };
    },
  };
  // @ts-ignore
  vscode.languages.registerInlineCompletionItemProvider({ pattern: '**' }, provider);
}
